flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/TableScanTest.scala

```diff
@@ -281,25 +281,40 @@ class TableScanTest extends TableTestBase {
 
   @Test
   def testJoinOnChangelogSource(): Unit = {
+    verifyJoinOnSource("I,UB,UA")
+  }
+
+  @Test
+  def testJoinOnNoUpdateSource(): Unit = {
+    verifyJoinOnSource("I,D")
+  }
+
+  @Test
+  def testJoinOnUpsertSource(): Unit = {
+    verifyJoinOnSource("UA,D")
+  }
+
+  private def verifyJoinOnSource(changelogMode: String): Unit = {
     util.addTable(
       """
-         |CREATE TABLE orders (
-         |  amount BIGINT,
-         |  currency STRING
-         |) WITH (
-         | 'connector' = 'values',
-         | 'changelog-mode' = 'I'
-         |)
-         |""".stripMargin)
+        |CREATE TABLE orders (
+        |  amount BIGINT,
+        |  currency STRING
+        |) WITH (
+        | 'connector' = 'values',
+        | 'changelog-mode' = 'I'
+        |)
+        |""".stripMargin)
     util.addTable(
-      """
-         |CREATE TABLE rates_history (
-         |  currency STRING,
-         |  rate BIGINT
-         |) WITH (
-         |  'connector' = 'values',
-         |  'changelog-mode' = 'I,UB,UA'
-         |)
+      s"""
+        |CREATE TABLE rates_history (
+        |  currency STRING,
+        |  rate BIGINT,
+        |  PRIMARY KEY (currency) NOT ENFORCED
```

Comment:
could you change the join key to `(currency, currency_no)` to check the join key contains unique key case.